File: src/core/git-manager.ts
import { open, readFile, stat } from "node:fs/promises";
import { execFile } from "node:child_process";
import { resolve, sep } from "node:path";
import { promisify } from "node:util";
import { assignLanes, type GitGraphLayoutRow } from "./git-graph-layout.js";

const execFileAsync = promisify(execFile);

export interface GitFileStatus {
  path: string;
  index: string;
  workingTree: string;
}

export interface GitStatus {
  branch: string;
  files: GitFileStatus[];
}

export interface GitLogEntry {
  hash: string;
  subject: string;
}

export type GitGraphRefKind = "head" | "branch" | "remote" | "tag";

export interface GitGraphRef {
  name: string;
  kind: GitGraphRefKind;
}

export interface GitGraphCommit {
  hash: string;
  parents: string[];
  author: string;
  email: string;
  timestamp: number;
  subject: string;
  refs: GitGraphRef[];
  lane: number;
  laneCount: number;
  edges: GitGraphLayoutRow["edges"];
  throughLanes: number[];
}

export interface GitBranch {
  name: string;
  current: boolean;
  remote: boolean;
  upstream?: string;
}

/** One tracked text file's size, for the "largest files" ranking. */
export interface FileSizeRank {
  path: string;
  /** Line count; a lower bound (`truncated: true`) for files past the read cap. */
  lines: number;
  bytes: number;
  truncated: boolean;
}

export class GitManager {
  constructor(private readonly cwd = process.cwd()) {}

  async status(): Promise<GitStatus> {
    const [branch, porcelain] = await Promise.all([
      this.git(["branch", "--show-current"]),
      this.git(["status", "--porcelain=v1", "--untracked-files=all"]),
    ]);

    return {
      branch: branch.trim(),
      files: porcelain
        .split("\n")
        .filter(Boolean)
        .map((line) => ({
          index: line[0] ?? " ",
          workingTree: line[1] ?? " ",
          path: line.slice(3),
        })),
    };
  }

  async diff(path?: string): Promise<string> {
    return this.git(path ? ["diff", "--", path] : ["diff"]);
  }

  async stagedDiff(path?: string): Promise<string> {
    return this.git(path ? ["diff", "--cached", "--", path] : ["diff", "--cached"]);
  }

  async fileDiff(file: GitFileStatus): Promise<string> {
    if (file.index === "?" && file.workingTree === "?") {
      return this.untrackedDiff(file.path);
    }

    if (file.workingTree.trim()) {
      return this.diff(file.path);
    }

    if (file.index.trim()) {
      return this.stagedDiff(file.path);
    }

    return this.diff(file.path);
  }

  async log(limit = 10): Promise<GitLogEntry[]> {
    const output = await this.git([
      "log",
      `-${limit}`,
      "--pretty=format:%H%x09%s",
    ]);

    return output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const [hash = "", subject = ""] = line.split("\t");
        return { hash, subject };
      });
  }

  async graph(limit = 200): Promise<GitGraphCommit[]> {
    const recordSep = "\x1e";
    const fieldSep = "\x1f";
    const format = ["%H", "%P", "%an", "%ae", "%at", "%s"].join(fieldSep) + recordSep;

    const [logOutput, refsOutput, headHash] = await Promise.all([
      this.git([
        "log",
        "--all",
        "--date-order",
        `--max-count=${limit}`,
        `--pretty=format:${format}`,
      ]),
      this.git([
        "for-each-ref",
        "--format=%(refname)%09%(objectname)",
        "refs/heads",
        "refs/remotes",
        "refs/tags",
      ]),
      this.git(["rev-parse", "HEAD"]).catch(() => ""),
    ]);

    const refsByHash = new Map<string, GitGraphRef[]>();
    for (const line of refsOutput.split("\n").filter(Boolean)) {
      const [refName = "", hash = ""] = line.split("\t");
      if (!refName || !hash) continue;
      if (refName.endsWith("/HEAD")) continue;
      let name = refName;
      let kind: GitGraphRefKind;
      if (refName.startsWith("refs/heads/")) {
        name = refName.slice("refs/heads/".length);
        kind = "branch";
      } else if (refName.startsWith("refs/remotes/")) {
        name = refName.slice("refs/remotes/".length);
        kind = "remote";
      } else if (refName.startsWith("refs/tags/")) {
        name = refName.slice("refs/tags/".length);
        kind = "tag";
      } else {
        continue;
      }
      const list = refsByHash.get(hash) ?? [];
      list.push({ name, kind });
      refsByHash.set(hash, list);
    }

    const trimmedHead = headHash.trim();
    if (trimmedHead) {
      const list = refsByHash.get(trimmedHead) ?? [];
      list.unshift({ name: "HEAD", kind: "head" });
      refsByHash.set(trimmedHead, list);
    }

    const parsed = logOutput
      .split(recordSep)
      .map((record) => record.replace(/^\n/, ""))
      .filter((record) => record.length > 0)
      .map((record) => {
        const [hash = "", parentField = "", author = "", email = "", timestamp = "0", subject = ""] =
          record.split(fieldSep);
        return {
          hash,
          parents: parentField ? parentField.split(" ").filter(Boolean) : [],
          author,
          email,
          timestamp: Number(timestamp) || 0,
          subject,
        };
      })
      .filter((commit) => commit.hash);

    const layout = assignLanes(parsed.map(({ hash, parents }) => ({ hash, parents })));

    return parsed.map((commit, index) => ({
      ...commit,
      refs: refsByHash.get(commit.hash) ?? [],
      lane: layout[index].lane,
      laneCount: layout[index].laneCount,
      edges: layout[index].edges,
      throughLanes: layout[index].throughLanes,
    }));
  }

  async listTrackedFiles(): Promise<string[]> {
    const output = await this.git(["ls-files", "-z"]);
    return output.split("\0").filter((path) => path.length > 0);
  }

  /**
   * Rank tracked files by line count, longest first — a maintainability lens.
   * Stays cheap: `ls-files` already excludes deps/build output, binaries are
   * skipped, reads are capped, and only a bounded number run concurrently.
   */
  async rankFilesBySize(
    options: { maxBytes?: number; limit?: number; concurrency?: number } = {},
  ): Promise<FileSizeRank[]> {
    const maxBytes = options.maxBytes ?? 2_000_000;
    const concurrency = options.concurrency ?? 16;
    const files = await this.listTrackedFiles();
    const ranks: FileSizeRank[] = [];

    let cursor = 0;
    const worker = async (): Promise<void> => {
      while (cursor < files.length) {
        const path = files[cursor++];
        const rank = await this.measureFile(path, maxBytes).catch(() => null);
        if (rank) ranks.push(rank);
      }
    };
    await Promise.all(
      Array.from({ length: Math.min(concurrency, files.length) }, worker),
    );

    ranks.sort((a, b) => b.lines - a.lines || b.bytes - a.bytes);
    return options.limit ? ranks.slice(0, options.limit) : ranks;
  }

  /** Count lines in one tracked file; returns null for binaries or non-files. */
  private async measureFile(
    path: string,
    maxBytes: number,
  ): Promise<FileSizeRank | null> {
    const fullPath = resolveInside(this.cwd, path);
    const info = await stat(fullPath);
    if (!info.isFile()) return null;

    const bytes = info.size;
    const truncated = bytes > maxBytes;
    let buffer: Buffer;
    if (truncated) {
      // Don't load a huge file just to rank it — the first slice is enough.
      const handle = await open(fullPath, "r");
      try {
        buffer = Buffer.alloc(maxBytes);
        await handle.read(buffer, 0, maxBytes, 0);
      } finally {
        await handle.close();
      }
    } else {
      buffer = await readFile(fullPath);
    }

    if (buffer.includes(0)) return null; // binary

    let lines = 0;
    for (let i = 0; i < buffer.length; i++) {
      if (buffer[i] === 10) lines++;
    }
    // Count a final line that has no trailing newline.
    if (buffer.length > 0 && buffer[buffer.length - 1] !== 10) lines++;

    return { path, lines, bytes, truncated };
  }

  async readTrackedFile(
    path: string,
    options: { maxBytes?: number } = {},
  ): Promise<{ content: string; truncated: boolean; binary: boolean; size: number }> {
    const maxBytes = options.maxBytes ?? 1_000_000;
    const tracked = new Set(await this.listTrackedFiles());
    if (!tracked.has(path)) {
      throw new Error("file is not tracked by git");
    }
    const fullPath = resolveInside(this.cwd, path);
    const buffer = await readFile(fullPath);
    const binary = buffer.includes(0);
    if (binary) {
      return { content: "", truncated: false, binary: true, size: buffer.length };
    }
    const truncated = buffer.length > maxBytes;
    const slice = truncated ? buffer.subarray(0, maxBytes) : buffer;
    return {
      content: slice.toString("utf8"),
      truncated,
      binary: false,
      size: buffer.length,
    };
  }

  async commitDiff(hash: string, path?: string): Promise<string> {
    const sha = this.validateHash(hash);
    const args = ["show", "--patch", "--format=", sha];
    if (path) {
      args.push("--", path);
    }
    // `git show` handles root commits (no parent) gracefully, unlike `diff <sha>^..<sha>`.
    return this.git(args);
  }

  async commitFiles(hash: string): Promise<GitFileStatus[]> {
    const sha = this.validateHash(hash);
    const output = await this.git([
      "show",
      "--name-status",
      "--format=",
      "-z",
      sha,
    ]);
    // With -z, fields are NUL-separated:
    //   regular:  STATUS \0 PATH \0
    //   rename:   R### \0 OLD \0 NEW \0   (also C### for copies)
    const files: GitFileStatus[] = [];
    const tokens = output.split("\0").filter((t) => t.length > 0);
    let i = 0;
    while (i < tokens.length) {
      const statusToken = tokens[i];
      const letter = (statusToken[0] ?? " ").toUpperCase();
      if (letter === "R" || letter === "C") {
        const newPath = tokens[i + 2];
        if (newPath) {
          files.push({ path: newPath, index: letter, workingTree: " " });
        }
        i += 3;
      } else {
        const path = tokens[i + 1];
        if (path) {
          files.push({ path, index: letter, workingTree: " " });
        }
        i += 2;
      }
    }
    return files;
  }

  private validateHash(hash: string): string {
    const sha = requireName(hash, "commit");
    if (!/^[0-9a-f]{4,64}$/i.test(sha)) {
      throw new Error("invalid commit hash");
    }
    return sha;
  }

  async branches(): Promise<GitBranch[]> {
    const output = await this.git([
      "branch",
      "--all",
      "--format=%(refname)%09%(HEAD)%09%(upstream:short)",
    ]);

    return output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const [refName = "", head = "", upstream = ""] = line.split("\t");
        const remote = refName.startsWith("refs/remotes/");
        const name = remote
          ? refName.slice("refs/remotes/".length)
          : refName.replace(/^refs\/heads\//, "");
        return {
          name,
          current: head === "*",
          remote,
          upstream: upstream || undefined,
        };
      })
      .filter((branch) => branch.name && !branch.name.endsWith("/HEAD"));
  }

  async switchBranch(branch: string): Promise<string> {
    const name = requireName(branch, "branch");
    const remoteBranches = (await this.branches()).filter((item) => item.remote);
    const isRemoteBranch = remoteBranches.some((item) => item.name === name);

    return this.git(isRemoteBranch ? ["switch", "--track", name] : ["switch", name]);
  }

  async createBranch(branch: string): Promise<string> {
    return this.git(["switch", "-c", requireName(branch, "branch")]);
  }

  async fetch(): Promise<string> {
    return this.git(["fetch", "--prune"]);
  }

  async stage(paths: string[]): Promise<string> {
    requirePaths(paths);
    return this.git(["add", "--", ...paths]);
  }

  async unstage(paths: string[]): Promise<string> {
    requirePaths(paths);
    return this.git(["restore", "--staged", "--", ...paths]);
  }

  async commit(message: string): Promise<string> {
    if (!message.trim()) {
      throw new Error("commit message is required");
    }

    return this.git(["commit", "-m", message]);
  }

  private async git(args: string[]): Promise<string> {
    try {
      const { stdout, stderr } = await execFileAsync("git", args, {
        cwd: this.cwd,
        // Data-file diffs/logs (CSV, logs) can exceed Node's 1 MB stdout
        // default, which otherwise aborts with a maxBuffer error whose partial
        // stdout gets surfaced to the UI as a raw, unrendered diff.
        maxBuffer: 256 * 1024 * 1024,
      });
      return stdout || stderr;
    } catch (error) {
      if (isExecError(error)) {
        throw new Error((error.stderr || error.stdout || error.message).trim());
      }
      throw error;
    }
  }

  private async untrackedDiff(path: string): Promise<string> {
    const fullPath = resolveInside(this.cwd, path);
    const content = await readFile(fullPath, "utf8");
    const lines = content.split("\n");
    const hasTrailingNewline = lines[lines.length - 1] === "";
    const contentLines = hasTrailingNewline ? lines.slice(0, -1) : lines;
    const hunkLength = Math.max(contentLines.length, 1);
    const diffLines = [
      `diff --git a/${path} b/${path}`,
      "new file mode 100644",
      "index 0000000..0000000",
      "--- /dev/null",
      `+++ b/${path}`,
      `@@ -0,0 +1,${hunkLength} @@`,
      ...contentLines.map((line) => `+${line}`),
    ];

    if (!hasTrailingNewline) {
      diffLines.push("\\ No newline at end of file");
    }

    return `${diffLines.join("\n")}\n`;
  }
}

function resolveInside(root: string, path: string): string {
  const rootPath = resolve(root);
  const targetPath = resolve(rootPath, path);
  if (targetPath !== rootPath && !targetPath.startsWith(`${rootPath}${sep}`)) {
    throw new Error("file path must stay inside the repository");
  }
  return targetPath;
}

function requireName(value: string, label: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error(`${label} is required`);
  }
  return trimmed;
}

function requirePaths(paths: string[]): void {
  if (paths.length === 0 || paths.some((path) => !path.trim())) {
    throw new Error("at least one file path is required");
  }
}

function isExecError(
  error: unknown,
): error is Error & { stdout?: string; stderr?: string } {
  return error instanceof Error;
}

File: src-tauri/src/core/git_manager.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::path::Path;
use tokio::process::Command;

// ---------------------------------------------------------------------------
// Output types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileStatus {
    pub path: String,
    pub index: String,
    pub working_tree: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStatus {
    pub branch: String,
    pub upstream: Option<String>,
    pub ahead: i32,
    pub behind: i32,
    pub files: Vec<GitFileStatus>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommit {
    pub hash: String,
    pub short_hash: String,
    pub subject: String,
    pub author: String,
    pub email: String,
    pub date: String,
    pub refs: Vec<String>,
    pub parents: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranch {
    pub name: String,
    pub is_current: bool,
    pub is_remote: bool,
}

// ---------------------------------------------------------------------------
// Runner helpers
// ---------------------------------------------------------------------------

async fn git(cwd: &str, args: &[&str]) -> Result<String> {
    let out = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .await
        .context("git command failed")?;
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

async fn git_lines(cwd: &str, args: &[&str]) -> Result<Vec<String>> {
    let raw = git(cwd, args).await?;
    Ok(raw.lines().map(str::to_string).filter(|l| !l.is_empty()).collect())
}

// ---------------------------------------------------------------------------
// GitManager
// ---------------------------------------------------------------------------

pub struct GitManager;

impl GitManager {
    pub async fn status(cwd: &str) -> Result<GitStatus> {
        // Branch name
        let branch = git(cwd, &["rev-parse", "--abbrev-ref", "HEAD"]).await?.trim().to_string();

        // Upstream tracking
        let upstream_raw = git(cwd, &[
            "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}",
        ]).await.unwrap_or_default();
        let upstream = upstream_raw.trim().to_string();
        let upstream = if upstream.is_empty() || upstream.starts_with("fatal") {
            None
        } else {
            Some(upstream)
        };

        // Ahead/behind
        let (ahead, behind) = if upstream.is_some() {
            let ab = git(cwd, &["rev-list", "--left-right", "--count", "HEAD...@{u}"])
                .await
                .unwrap_or_default();
            let parts: Vec<i32> = ab.trim().split_whitespace()
                .filter_map(|s| s.parse().ok())
                .collect();
            (parts.first().copied().unwrap_or(0), parts.get(1).copied().unwrap_or(0))
        } else {
            (0, 0)
        };

        // File statuses (porcelain v1)
        let status_out = git(cwd, &["status", "--porcelain", "-u"]).await?;
        let files = status_out.lines()
            .filter(|l| l.len() >= 3)
            .map(|l| {
                let index = l.chars().next().map(|c| c.to_string()).unwrap_or_default();
                let wt = l.chars().nth(1).map(|c| c.to_string()).unwrap_or_default();
                let path = l[3..].trim().to_string();
                GitFileStatus { path, index, working_tree: wt }
            })
            .collect();

        Ok(GitStatus { branch, upstream, ahead, behind, files })
    }

    pub async fn diff(cwd: &str, file: Option<&str>) -> Result<String> {
        let mut args = vec!["diff", "--"];
        if let Some(f) = file {
            args.push(f);
        }
        let staged = git(cwd, &["diff", "--cached", "--"]).await?;
        let unstaged = git(cwd, &args).await?;
        if !staged.is_empty() { Ok(staged) } else { Ok(unstaged) }
    }

    pub async fn file_diff(cwd: &str, file: &str) -> Result<String> {
        // Try staged diff first, fall back to unstaged
        let staged = git(cwd, &["diff", "--cached", "--", file]).await?;
        if !staged.trim().is_empty() {
            return Ok(staged);
        }
        git(cwd, &["diff", "--", file]).await
    }

    pub async fn list_tracked_files(cwd: &str) -> Result<Vec<String>> {
        git_lines(cwd, &["ls-files"]).await
    }

    pub async fn read_file(cwd: &str, path: &str) -> Result<String> {
        tokio::fs::read_to_string(Path::new(cwd).join(path))
            .await
            .context("Failed to read tracked file")
    }

    pub async fn write_file(cwd: &str, path: &str, content: &str) -> Result<()> {
        let full = Path::new(cwd).join(path);
        tokio::fs::write(full, content).await.context("Failed to write file")
    }

    pub async fn graph(cwd: &str, limit: usize) -> Result<Vec<GitCommit>> {
        let fmt = "%H\x1f%h\x1f%s\x1f%an\x1f%ae\x1f%aI\x1f%D\x1f%P";
        let limit_str = limit.to_string();
        let raw = git(cwd, &["log", &format!("-{limit_str}"), &format!("--format={fmt}")]).await?;

        let commits = raw.lines().filter(|l| !l.is_empty()).map(|line| {
            let parts: Vec<&str> = line.split('\x1f').collect();
            GitCommit {
                hash: parts.first().copied().unwrap_or("").to_string(),
                short_hash: parts.get(1).copied().unwrap_or("").to_string(),
                subject: parts.get(2).copied().unwrap_or("").to_string(),
                author: parts.get(3).copied().unwrap_or("").to_string(),
                email: parts.get(4).copied().unwrap_or("").to_string(),
                date: parts.get(5).copied().unwrap_or("").to_string(),
                refs: parts.get(6).copied().unwrap_or("").split(", ")
                    .filter(|s| !s.is_empty())
                    .map(str::to_string)
                    .collect(),
                parents: parts.get(7).copied().unwrap_or("").split_whitespace()
                    .map(str::to_string)
                    .collect(),
            }
        }).collect();

        Ok(commits)
    }

    pub async fn commit_diff(cwd: &str, hash: &str, file: Option<&str>) -> Result<String> {
        let mut args = vec!["show", hash, "--"];
        if let Some(f) = file { args.push(f); }
        git(cwd, &args).await
    }

    pub async fn commit_files(cwd: &str, hash: &str) -> Result<Vec<String>> {
        git_lines(cwd, &["diff-tree", "--no-commit-id", "-r", "--name-only", hash]).await
    }

    pub async fn stage(cwd: &str, paths: &[String]) -> Result<()> {
        let mut args = vec!["add", "--"];
        let path_refs: Vec<&str> = paths.iter().map(String::as_str).collect();
        args.extend(path_refs);
        let out = Command::new("git").args(&args).current_dir(cwd).output().await?;
        if !out.status.success() {
            anyhow::bail!("{}", String::from_utf8_lossy(&out.stderr));
        }
        Ok(())
    }

    pub async fn unstage(cwd: &str, paths: &[String]) -> Result<()> {
        let mut args = vec!["restore", "--staged", "--"];
        let path_refs: Vec<&str> = paths.iter().map(String::as_str).collect();
        args.extend(path_refs);
        Command::new("git").args(&args).current_dir(cwd).output().await?;
        Ok(())
    }

    pub async fn commit(cwd: &str, message: &str) -> Result<String> {
        let out = Command::new("git")
            .args(["commit", "-m", message])
            .current_dir(cwd)
            .output()
            .await?;
        if !out.status.success() {
            anyhow::bail!("{}", String::from_utf8_lossy(&out.stderr));
        }
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }

    pub async fn push(cwd: &str, remote: Option<&str>) -> Result<String> {
        let remote = remote.unwrap_or("origin");
        let out = Command::new("git")
            .args(["push", remote])
            .current_dir(cwd)
            .output()
            .await?;
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }

    pub async fn fetch(cwd: &str) -> Result<String> {
        let out = Command::new("git").args(["fetch"]).current_dir(cwd).output().await?;
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }

    pub async fn create_branch(cwd: &str, name: &str) -> Result<()> {
        let out = Command::new("git")
            .args(["checkout", "-b", name])
            .current_dir(cwd)
            .output()
            .await?;
        if !out.status.success() {
            anyhow::bail!("{}", String::from_utf8_lossy(&out.stderr));
        }
        Ok(())
    }

    pub async fn switch_branch(cwd: &str, name: &str) -> Result<()> {
        let out = Command::new("git")
            .args(["checkout", name])
            .current_dir(cwd)
            .output()
            .await?;
        if !out.status.success() {
            anyhow::bail!("{}", String::from_utf8_lossy(&out.stderr));
        }
        Ok(())
    }

    pub async fn branches(cwd: &str) -> Result<Vec<GitBranch>> {
        let current = git(cwd, &["rev-parse", "--abbrev-ref", "HEAD"])
            .await?.trim().to_string();
        let locals = git_lines(cwd, &["branch", "--format=%(refname:short)"]).await?;
        let remotes = git_lines(cwd, &["branch", "-r", "--format=%(refname:short)"]).await?;

        let mut branches: Vec<GitBranch> = locals.into_iter().map(|name| GitBranch {
            is_current: name == current,
            is_remote: false,
            name,
        }).collect();

        branches.extend(remotes.into_iter().map(|name| GitBranch {
            is_current: false,
            is_remote: true,
            name,
        }));

        Ok(branches)
    }

    pub async fn pull_default(cwd: &str) -> Result<String> {
        // Checkout default branch (main or master) and pull --ff-only
        let default = detect_default_branch(cwd).await;
        Command::new("git").args(["checkout", &default]).current_dir(cwd).output().await?;
        let out = Command::new("git").args(["pull", "--ff-only"]).current_dir(cwd).output().await?;
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }
}

async fn detect_default_branch(cwd: &str) -> String {
    let out = Command::new("git")
        .args(["symbolic-ref", "refs/remotes/origin/HEAD"])
        .current_dir(cwd)
        .output()
        .await
        .unwrap_or_default();
    let raw = String::from_utf8_lossy(&out.stdout);
    raw.trim().split('/').last().unwrap_or("main").to_string()
}
